### PR TITLE
qcvar handling on wfn, take 2

### PR DESCRIPTION
**Pull request by loriab**
## Description
Suceeds #1381 as a step toward #883 by improving qcvar capabilities on Wfn and localizing some modules.

## Todos
Notable points (developer or user-interest) that this PR has or will accomplish.
- [x] improves the wfn localization of qcvars for DFT, scf_helper, DFMP2, OEPROP. trying to set on wfn c-side, then just copy into P::e.globals in proc.py.
- [x] moves DETCI, DFMP2 entirely to qcvar passing on Wfn. OCC, DFOCC, FNOCC do so for mp2-like only.
- [x] adds some functional-including qcvars for DFT
- [x] BREAKS
  - remove `SCS-MP2 OPPOSITE-SPIN CORRELATION ENERGY` and `SCS-MP2 SAME-SPIN CORRELATION ENERGY` from dfmp2 `Wavefunction.variables`
  - remove MP2 psivars from double-hybrid DFT final return
  - `SCS-MP2 * ENERGY` now are defined (6/5, 1/3) quantities. For the user-influenceable vars, it's `CUSTOM SCS-MP2 * ENERGY`

## Notes
- reference E for detci https://github.com/psi4/psi4/pull/1381#discussion_r237680985

## Checklist
- [ ] Tests added for any new features
- [ ] [All or relevant fraction of full tests run](http://psicode.org/psi4manual/master/build_planning.html#how-to-run-a-subset-of-tests)

## Status
- [x] Ready for review
- [x] Ready for merge

### CODE REVIEW

File: psi4/src/psi4/occ/manager.cc [hunk 1]
```diff
@@ -214,17 +214,19 @@ void OCCWave::omp2_manager() {
...
         Process::environment.globals["SCSN-OMP2 TOTAL ENERGY"] = Escsnmp2;
         Process::environment.globals["SCS-OMP2-VDW TOTAL ENERGY"] = Escsmp2vdw;
         Process::environment.globals["SOS-PI-OMP2 TOTAL ENERGY"] = Esospimp2;
-        Process::environment.globals["CURRENT ENERGY"] = Emp2L;
-        Process::environment.globals["CURRENT REFERENCE ENERGY"] = Escf;
-        Process::environment.globals["CURRENT CORRELATION ENERGY"] = Emp2L - Escf;
+        energy_ = Emp2L;
+        variables_["CURRENT ENERGY"] = Emp2L;
+        variables_["CURRENT REFERENCE ENERGY"] = Escf;
+        variables_["CURRENT CORRELATION ENERGY"] = Emp2L - Escf;
```
[the remark this answers is not in the document]
Comment (loriab) [reply psi4/src/psi4/occ/manager.cc]: Basically vatiables_ and energy_ are what I vouch for and test. The globals are what I don’t feel like breaking or getting into an argument about yet. Ultimately they’ll be handled by the qcvar identity formulas py-side from the wfn bars set c-side.

File: psi4/src/psi4/detci/diag_h.cc [hunk 2]
```diff
@@ -472,32 +472,33 @@ int CIWavefunction::diag_h(double conv_e, double conv_rms) {
     // Write out energy to psivars
     tval = evals[Parameters_->root] + edrc + nucrep;
 
-    Process::environment.globals["CURRENT ENERGY"] = tval;
-    Process::environment.globals["CURRENT CORRELATION ENERGY"] = tval - CalcInfo_->escf;
-    Process::environment.globals["CURRENT REFERENCE ENERGY"] = CalcInfo_->escf;
-    Process::environment.globals["CI TOTAL ENERGY"] = tval;
-    Process::environment.globals["CI CORRELATION ENERGY"] = tval - CalcInfo_->escf;
+    set_energy(tval);
+    set_scalar_variable("CURRENT ENERGY", tval);
```
[the remark this answers is not in the document]
Comment (loriab) [reply psi4/src/psi4/detci/diag_h.cc]: Yes, wfn.energy_, and global and wfn current energy should all be redundant. You’re probably right — I should check the slack discussion as to whether the slots on wfn are independent. This code predates those decisions.

[the remark this answers is not in the document]
Comment (loriab) [reply psi4/src/psi4/detci/diag_h.cc]: Added [roadmap](https://github.com/psi4/psi4/issues/883#issuecomment-449694968) and first bit of data overloading, which I think is safe enough b/c wfn vars underused.

**Comment by loriab** [comment]
This has been rebased, and now passes all tests. https://github.com/psi4/psi4/pull/1445#issuecomment-452033368 is extinct. Ready for review.

File: psi4/driver/procrouting/proc.py [hunk 3]
```diff
@@ -1904,6 +1919,11 @@ def run_occ(name, **kwargs):
 
     occ_wfn = core.occ(ref_wfn)
 
+    # Shove variables into global space
+    if name in ['mp2', 'omp2', 'scs-omp2', 'scs(n)-omp2', 'scs-omp2-vdw', 'sos-omp2']:
```
Comment (dgasmith) [inline psi4/driver/procrouting/proc.py]: Can this be simplified here and at other similar lines? Seems a bit brittle and easy to miss in the future.

Comment (loriab) [reply psi4/driver/procrouting/proc.py]: It's all the names in the mp2 if/elif block above. No way to accumulate them, really unless `'mp2' in name and not name.endswith('mp2.5')`. only way I see would be to break `run_occ` up into mp2/mp2.5/mp3/ccd `run_` fns. while I wholly agree that it's brittle (along with all of occ), it should at least heal toward removing that `if name in [` line as further wfn psivar passes hit mp3, cc, etc.

other similar lines?

Comment (dgasmith) [reply psi4/driver/procrouting/proc.py]: Could we have a boolean that is flip above then? Might be less dangerous if we localize the variable change.

**Pull request merged**